File: 2015/13/13.py
from math      import sqrt
from itertools import permutations, pairwise


def parser(input_):
    changes = list()
    for change in input_.splitlines():
        gain = int(change.split(' ')[3])
        changes.append(gain if 'gain' in change else -gain)
    return changes


def solver(changes):
    
    def compute_changes(myself):
        n    = 1 + int(sqrt(len(changes))) + myself
        best = 0
        for arrangement in permutations(range(n)):
            acc = 0
            for (a, b) in pairwise(arrangement + (arrangement[0],)):
                if n - myself in (a, b): continue
                i1 = (n - 1 - myself) * a + b - (b > a)
                i2 = (n - 1 - myself) * b + a - (a > b)
                acc    += changes[i1] + changes[i2]
            if acc > best: best = acc
        return best

    yield compute_changes(myself = False)
    yield compute_changes(myself = True)

File: 2015/13/13.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<i16> {
    input
        .lines()
        .map(|l| l.split_whitespace().collect_vec())
        .map(|v| match v[2] {
            "gain" => v[3].parse::<i16>().unwrap(),
            "lose" => -v[3].parse::<i16>().unwrap(),
            _ => unreachable!(),
        })
        .collect_vec()
}


fn part_1(changes: Vec<i16>) -> i16 {
    solve(changes,1)
    }

fn part_2(changes: Vec<i16>) -> i16 {
    solve(changes,2)
    }

fn solve(changes: Vec<i16>, part: usize) -> i16 {
    let c = ((changes.len()) as f64).sqrt() as usize + part ;
    (0..c)
    .permutations(c)
    .into_iter()
    .map(|perm| {
        perm
        .iter()
        .cycle()
        .take(c+1)
        .into_iter()
        .tuple_windows()
        .fold(0, |acc, (&a,&b)| {
            if ![a,b].contains(&(c - part + 1))  {
                let index1 = (c-part) * a + b - (b > a) as usize;
                let index2 = (c-part) * b + a - (a > b) as usize;
                acc + changes[index1] + changes[index2]
            } else {
                acc
            }})})
    .max()
    .unwrap()
}
